packages/uploadthing/src/effect-http.ts

```diff
@@ -0,0 +1,114 @@
+  buildRequestHandler,
+} from "./internal/handler";
+import { incompatibleNodeGuard } from "./internal/incompat-node-guard";
+import { toWebRequest } from "./internal/to-web-request";
+import type { FileRouter, RouteHandlerOptions } from "./internal/types";
+import type { CreateBuilderOptions } from "./internal/upload-builder";
+import { createBuilder } from "./internal/upload-builder";
+
+export { UTFiles } from "./internal/types";
+export type { FileRouter };
+
+type MiddlewareArgs = {
+  req: Http.request.ServerRequest;
+  res: undefined;
+  event: undefined;
+};
+
+export const createUploadthing = <TErrorShape extends Json>(
+  opts?: CreateBuilderOptions<TErrorShape>,
+) => createBuilder<MiddlewareArgs, TErrorShape>(opts);
+
+export const createRouteHandler = <TRouter extends FileRouter>(
+  opts: RouteHandlerOptions<TRouter>,
+): Http.router.Router<Http.body.BodyError, never> => {
+  incompatibleNodeGuard();
+
+  const requestHandler = buildRequestHandler<TRouter, MiddlewareArgs>(
+    opts,
+    "effect-http",
+  );
+  const getBuildPerms = buildPermissionsInfoHandler<TRouter>(opts);
+
+  return Http.router.empty.pipe(
+    Http.router.get(
+      "/",
+      Http.response.json(getBuildPerms(), {
+        headers: Http.headers.fromInput({
+          "x-uploadthing-version": UPLOADTHING_VERSION,
+        }),
+      }),
+    ),
+    Http.router.post(
+      "/",
+      Effect.flatMap(Http.request.ServerRequest, (req) =>
+        requestHandler({
+          /**
+           * TODO: Redo this to be more cross-platform
+           * This should handle WinterCG and Node.js runtimes,
+           * unsure about others...
+           * Perhaps we can use `Http.request.ServerRequest` internally?
+           */
+          req: Effect.if(req.source instanceof Request, {
+            onTrue: () => Effect.succeed(req.source as Request),
+            onFalse: () =>
+              req.json.pipe(
+                // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
+                Effect.flatMap((body) => toWebRequest(req.source as any, body)),
+                Effect.catchTag(
+                  "RequestError",
+                  (error) =>
+                    new UploadThingError({
+                      code: "BAD_REQUEST",
+                      message: "INVALID_JSON",
+                      cause: error,
+                    }),
+                ),
+              ),
+          }),
+          middlewareArgs: { req, res: undefined, event: undefined },
+        }).pipe(
+          Effect.provideService(FetchContext, {
+            fetch: opts.config?.fetch ?? globalThis.fetch,
+            baseHeaders: {
+              "x-uploadthing-version": UPLOADTHING_VERSION,
+              // These are filled in later in `parseAndValidateRequest`
+              "x-uploadthing-api-key": undefined,
+              "x-uploadthing-be-adapter": undefined,
+              "x-uploadthing-fe-package": undefined,
+            },
+          }),
```

Comment:
continuing from our `Client` implementation:
```ts
Effect.provideServiceEffect(
  HttpClient.Client,
  HttpClient.mapRequest(
    HttpClient.fetch,
    ClientRequest.setHeader("x-uploadthing-version", UPLOADTHING_VERSION),
  ),
),
Effect.provideService(HttpClient.Fetch, opts.config?.fetch ?? globalThis.fetch),
```


Note though, currently this means that you must use `fetch`, `Undici` and `Agent`s are not supported. But you might want to support them

Reply:
cool - the client stuff is on a separate branch so i'll add that in there.

what specifically is it that doesn't work with other fetch clients?